units/en/unit2/smolagents/multi_agent_systems.md

```diff
@@ -0,0 +1,316 @@
+    # Add 1 hour for takeoff and landing procedures
+    flight_time = (actual_distance / cruising_speed_kmh) + 1.0
+
+    # Format the results
+    return flight_time
+
+
+print(calculate_cargo_travel_time((41.8781, -87.6298), (-33.8688, 151.2093)))import os
+```
+
+```python
+from PIL import Image
+from smolagents import CodeAgent, DuckDuckGoSearchTool, OpenAIServerModel, VisitWebpageTool
+
+model = OpenAIServerModel(model_id="gpt-4o")
+```
+
+We can start with creating a baseline, simple agent to give us a simple report.
+
+```python
+task = """Find all Batman filming locations in the world, calculate the time to transfer via cargo plane to here (we're in Gotham, 40.7128° N, 74.0060° W), and return them to me as a pandas dataframe.
+Also give me some supercar factories with the same cargo plane transfer time."""
+```
+
+```python
+import pandas as pd
+
+agent = CodeAgent(
+    model=model,
+    tools=[DuckDuckGoSearchTool(), VisitWebpageTool(), calculate_cargo_travel_time],
+    additional_authorized_imports=["pandas"],
+    max_steps=20,
+)
+
+result = agent.run(task)result
+```
+
+We could already improve this a bit by throwing in some dedicated planning steps, and adding more prompting.
+
+```python
+agent.planning_interval = 4
+
+detailed_report = agent.run(f"""
+You're an expert analyst. You make comprehensive reports after visiting many websites.
+Don't hesitate to search for many queries at once in a for loop.
+For each data point that you find, visit the source url to confirm numbers.
+
+{task}
+""")
+
+print(detailed_report)
+```
+
+```python
+detailed_report
+```
+
+Thanks to these quick changes, we obtained a much more concise report by simply providing our agent a detailed prompt, and giving it planning capabilities!
+
+💸 But as you can see, the context window is quickly filling up. So **if we ask our agent to combine the results of detailed search with another, it will be slower and quickly ramp up tokens and costs**.
+
+➡️ We need to improve the structure of our system.
+## ✌️ Splitting the task between two agents
+
+Multi-agent structures allow to separate memories between different sub-tasks, with two great benefits:
+- Each agent is more focused on its core task, thus more performant
+- Separating memories reduces the count of input tokens at each step, thus reducing latency and cost.
+
+Let's create a team with a dedicated web search agent, managed by another agent.
+
+The manager agent should have plotting capabilities to redact its final report: so let us give it access to additional imports, including `matplotlib`, and `geopandas` + `shapely` for spatial plotting.
+
+```python
+!pip install matplotlib geopandas shapely -q
+
+model = OpenAIServerModel("gpt-4o")
+
+web_agent = CodeAgent(
+    model=model,
+    tools=[DuckDuckGoSearchTool(), VisitWebpageTool(), calculate_cargo_travel_time],
+    name="web_agent",
+    description="Browses the web to find information",
+    verbosity_level=1,
+    max_steps=10,
+)
+```
```

Comment:
```suggestion
Let's first install the required packages.

```python
!pip install matplotlib geopandas shapely -q
```

Now, we can start initialise our agent.

```python
model = OpenAIServerModel("gpt-4o")

web_agent = CodeAgent(
    model=model,
    tools=[DuckDuckGoSearchTool(), VisitWebpageTool(), calculate_cargo_travel_time],
    name="web_agent",
    description="Browses the web to find information",
    verbosity_level=1,
    max_steps=10,
)
```
```

Reply:
Sshould we add more import here too?